javalib/src/main/scala/java/util/AbstractList.scala

```diff
@@ -118,54 +125,61 @@ abstract class AbstractList[E] protected () extends AbstractCollection[E]
     if (index < 0 || index > size)
       throw new IndexOutOfBoundsException(index.toString)
   }
+}
 
-  private abstract class AbstractListView(fromIndex: Int,
-      private var toIndex: Int) extends AbstractList[E] {
+private abstract class AbstractListView[E](protected val list: List[E],
+    fromIndex: Int, protected var toIndex: Int) extends AbstractList[E] {
 
-    override def add(index: Int, e: E): Unit = {
-      checkIndexOnBounds(index)
-      self.add(fromIndex + index, e)
-      toIndex += 1
-    }
+  override def add(index: Int, e: E): Unit = {
+    checkIndexOnBounds(index)
+    list.add(fromIndex + index, e)
+    changeViewSize(1)
+  }
 
-    override def addAll(index: Int, c: Collection[_ <: E]): Boolean = {
-      checkIndexOnBounds(index)
-      self.addAll(fromIndex + index, c)
-      val elementsAdded = c.size
-      toIndex += elementsAdded
-      elementsAdded != 0
-    }
+  override def addAll(index: Int, c: Collection[_ <: E]): Boolean = {
+    checkIndexOnBounds(index)
+    list.addAll(fromIndex + index, c)
+    val elementsAdded = c.size
+    toIndex += elementsAdded
+    elementsAdded != 0
+  }
 
-    override def addAll(c: Collection[_ <: E]): Boolean =
-      addAll(size, c)
+  override def addAll(c: Collection[_ <: E]): Boolean =
+    addAll(size, c)
 
-    def get(index: Int): E = {
-      checkIndexInBounds(index)
-      self.get(fromIndex + index)
-    }
+  def get(index: Int): E = {
+    checkIndexInBounds(index)
+    list.get(fromIndex + index)
+  }
 
-    override def remove(index: Int): E = {
-      checkIndexInBounds(index)
-      val elem = self.remove(fromIndex + index)
-      toIndex -= 1
-      elem
-    }
+  override def remove(index: Int): E = {
+    checkIndexInBounds(index)
+    val elem = list.remove(fromIndex + index)
+    changeViewSize(-1)
+    elem
+  }
 
-    override def set(index: Int, e: E): E = {
-      checkIndexInBounds(index)
-      self.set(fromIndex + index, e)
-    }
+  override def set(index: Int, e: E): E = {
+    checkIndexInBounds(index)
+    list.set(fromIndex + index, e)
+  }
 
-    def size(): Int =
-      toIndex - fromIndex
+  def size(): Int =
+    toIndex - fromIndex
 
-    override protected def removeRange(_fromIndex: Int, _toIndex: Int): Unit = {
-      super.removeRange(_fromIndex, _toIndex)
-      toIndex -= _toIndex - _fromIndex
-    }
+  protected def changeViewSize(delta: Int): Unit =
+    toIndex += delta
+}
+
+private trait SizeChangeEvent {
+  protected var end: Int
 
+  protected def changeSize(delta: Int): Unit = {
```

Comment:
It would be good to mark this `@inline` to compensate for the indirections caused by the mixin.